Update apt repos and fix MySQL

jasonaowen opened the pull request:
A few weeks ago, [Oracle started signing their MySQL apt repo with a new key](https://bugs.mysql.com/bug.php?id=106188), partially in response to [their old key expiring](https://bugs.mysql.com/bug.php?id=105632). While we only use the MySQL repo in production for the `mysql` command line client (and that only for the convenience of devs), the key rotation broke the ability for devs to build new working Vagrant VMs.

So, since we need to mess with signing keys and apt repos, let's do it right. @slifty has previously done some similar work in https://github.com/OpenTechStrategies/torque-devenv/pull/29, and this builds upon what he learned there.

Reformat our third-party repo listings as [DEB822 multiline files](https://repolib.readthedocs.io/en/latest/deb822-format.html), and then pin our signing keys - including the new MySQL repo key. See additional details in the individual commit messages.

# Testing

I plan to test this branch by building new images soon, and will coordinate with @cecilia-donnelly to make sure not to step on her toes this week.

It would also be very helpful to build a new Vagrant VM; I did the thing where I put somewhat-difficult-to-reproduce-state into my instance, so I am slightly hesitant - @slifty I think I saw you comment on another PR that you don't have a working local right now, so if you have bandwidth it might be good for you to rebuild your devenv? Otherwise I will do later.

# Deployment

Deploying this does require a full deploy - building new images and running terraform. The upside to doing so is that we can use `mysql`... although I did manually update the key in prod and dev, to get something done and to test some changes, respectively. So this is probably a low-priority deployment, and depending on engineering bandwidth may need to wait to go out.

slifty commented:
Great idea (two birds + one stone ;) )

I'll spend my next cycle on testing this.

slifty commented:
I definitely lied about spending my next cycle on this BUT hey I'm here now.

I'm getting an issue:

```
Package 'mysql-server' has no installation candidate
```

This is related to the devenv's `deploy.sh`, however, so likely outside of the scope of this PR.  I'm only mentioning it here because it's, ironically, related to mysql.  I'll open an issue in devenv.

jasonaowen commented:
@slifty next debugging steps:

1. are there any other error messages before that line?
2. are the apt repo files getting copied into `/etc/apt/sources.list.d`?
3. do the other repos work - New Relic and nodesource? `apt policy nodejs` should show versions with `nodesource` in the version number, and `apt search newrelic` should show a handful of packages
4. do you get any warnings or errors during `sudo apt update`?

slifty commented on provisioners/configure.sh:
```diff
@@ -22,14 +22,8 @@ echo newrelic-php5 newrelic-php5/application-name string $NEW_RELIC_APPLICATION_
-# Add custom sources
-cp $TEMPLATES_PATH/etc/apt/sources.list.d/* /etc/apt/sources.list.d/
+cp $TEMPLATES_PATH/usr/share/keyrings/*.gpg /usr/share/keyrings/
+cp $TEMPLATES_PATH/etc/apt/sources.list.d/*.list /etc/apt/sources.list.d/
```
I believe this should be:

```
cp $TEMPLATES_PATH/etc/apt/sources.list.d/*.sources /etc/apt/sources.list.d/
```

(`.*list` => `.*sources`)

jasonaowen replied:
Oops, yes, you're absolutely right!

slifty commented:
Yes!  I unfortunately had to step away before actually doing much debugging -- I'll be posting some feedback via review (that way we have threads!)

slifty commented on templates/usr/share/keyrings/mysql-archive-keyring-2022.gpg:
```diff
@@ -0,0 +1,52 @@
+-----BEGIN PGP PUBLIC KEY BLOCK-----
```
I'm getting the following error from debian when I try to add new sources, which indicates that maybe these keys aren't being registered properly:

```
    permanent: W: GPG error: http://apt.newrelic.com/debian newrelic Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY B31B29E5548C16BF
    permanent: E: The repository 'http://apt.newrelic.com/debian newrelic Release' is not signed.
    permanent: W: GPG error: https://deb.nodesource.com/node_12.x buster InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 1655A0AB68576280
    permanent: E: The repository 'https://deb.nodesource.com/node_12.x buster InRelease' is not signed.
    permanent: W: GPG error: http://repo.mysql.com/apt/debian buster InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 467B942D3A79BD29
    permanent: E: The repository 'http://repo.mysql.com/apt/debian buster InRelease' is not signed.
```

Note that the files *are* copied; when I cat them they match the template, but when I cat other files that were that folder (e.g. `debian-archive-buster-security-automatic.gpg` it looks like they are binary files.  Indeed the gpg keys [in the commit you referenced as inspiration](https://github.com/OpenTechStrategies/torque-devenv/pull/29/commits/6e01a7e51818ec60e538124b2bdea0c673245895) for this change the gpg keys are binary files there too.

I think you missed the last "dearmor" step described in that commit message:

`cat ansible.asc | gpg --dearmor > ansible.gpg` <-- so if you run that on the gpg files it should do the trick

slifty replied:
I can confirm that running:

```
mv mysql-archive-keyring-2022.gpg mysql-archive-keyring-2022.asc
cat mysql-archive-keyring-2022.asc | gpg --dearmor > mysql-archive-keyring-2022.gpg
mv newrelic-archive-keyring.gpg newrelic-archive-keyring.asc
cat newrelic-archive-keyring.asc | gpg --dearmor > newrelic-archive-keyring.gpg
mv newrelic-archive-keyring.gpg newrelic-archive-keyring.asc
cat nodesource-archive-keyring.asc | gpg --dearmor > nodesource-archive-keyring.gpg
```

Solved the problem!  (we don't need the `.asc` files I just didn't want to give destructive advice in this message)

jasonaowen replied:
So this is very interesting!

I did intentionally omit the dearmor step, for two reasons: to allow us to easily compare with upstream, which all distribute armored keys; and because binary files and git are not close friends.

What's strange is that I have tested the ability to use armored keys with apt, and it's worked for me - both on my laptop running Ubuntu 20.04, and on the dev back-end EC2 instance.

The [Debian Wiki UseThirdParty article](https://wiki.debian.org/DebianRepository/UseThirdParty) has this quote:

> The reason why we avoid ASCII-armored files is that they can only be used by [SecureApt](https://wiki.debian.org/SecureApt) in version 1.4 or later (which appeared in stretch).

The stretch release is currently also known as `oldoldstable`, and here we're using buster aka `oldstable`. I was under the impression this was simply outdated advice, intended for backwards compatibility with a version we no longer care about. So I am very confused!

I would love to understand why they are working for me and are not working for you. Short of that, I think my preferred solution would be to dearmor at install time, rather than committing the binary keys - any concern about a bash `for` loop that does some ever-so-slightly-cleverness with `basename`?

jasonaowen replied:
I bit the bullet and recreated my vagrant VM with the key copy command corrected, and it worked for me. I'll push my changes, and then ask for you to re-test!

slifty replied:
Weird -- I wonder why the armored don't work for me.  (I did run it with the key copy correction and it didn't work until I de-armored)

I'm fine with the de-armoring happening as part of the provisioning (or if it has to be done manually as an additional step so it goes, though we should be careful not to add TOO many steps); so long as `vagrant up` works I'm fine!

slifty replied:
Pulled the latest -- the `.gpg` extension was the issue!  Now that they are `.asc` it works just fine as is (in fact, I may update the torque devenv to use the armored values for the reasons you described now that I see why it was trouble before!)

Thanks Jason!

jasonaowen replied:
Thanks for your help!

slifty reviewed (changes requested):
Thank you for this -- I can confirm that things won't run without them.

Two fixes to the PR are needed for merge:

1. The keys need to be dearmored.
2. The source file glob has to be modified to match the file names.

I believe that will fix it!  (I'm doing the dearmoring locally now to confirm, I'll comment in the appropriate thread if there was more to it than that)

jasonaowen commented:
@slifty I updated with the copy fix, and renamed the `*.gpg` keys to `*.asc`, which they should have been in the first place. I was able to do a `vagrant destroy && vagrant up` successfully, so please re-test and let me know if using armored keys is still broken for you!

slifty reviewed (approved):
The change to the copy script and rename from `gpg` to `asc` did the trick!

Noting here as well that `.gpg` actually did NOT work, I guess Debian cares what the extension is and treats the file differently depending.

jasonaowen commented:
> Noting here as well that .gpg actually did NOT work, I guess Debian cares what the extension is and treats the file differently depending.

Oh wild! I can confirm that: `sudo cp /usr/share/keyrings/nodesource-archive-keyring.{asc,gpg}; sudoedit /etc/apt/sources.list.d/nodesource.sources; sudo apt update` leads to a `NO_PUBKEY` error, and changing the signed-by line back to `.asc` fixes it. Neat! Very glad to figure that out. Thanks, @slifty!
